Windows line endings break escaped newlines in strings

The following does not work as expected if your file's line endings are `\r\n`:

```crystal
text = "This will have \
    newlines and \
    leading whitespace \
    if your line endings are \\r\\n."
```

Perhaps Crystal wants to support CRLF as a regular newline, perhaps it doesn't - whatever the case, I figured I'd report it.

I have an issue with CRLF.

- Installed Crystal 1.3.0 Windows
- Create a test app
- Add ameba to `development_dependencies`
- `shards install`

```
shards build
Dependencies are satisfied
Building: ameba
Error target ameba failed to compile:
In src\ameba\cli\cmd.cr:173:13

 173 | puts \
             ^
Error: unknown token: '\r'

make: *** [Makefile:8: bin/ameba] Error 1
```

Windows will inevitably create a file with CRLF somewhere, this is a reason why Crystal has to handle CRLF.

I continue my deep dive of Crystal with Windows 🌊🌊🌊

--- a/spec/compiler/parser/parser_spec.cr
+++ b/spec/compiler/parser/parser_spec.cr
@@ -1528,7 +1528,10 @@ module Crystal
 
     it_parses %("hello " \\\n "world"), StringLiteral.new("hello world")
     it_parses %("hello "\\\n"world"), StringLiteral.new("hello world")
+    it_parses %("hello " \\\r\n "world"), StringLiteral.new("hello world")
+    it_parses %("hello "\\\r\n"world"), StringLiteral.new("hello world")
     it_parses %("hello \#{1}" \\\n "\#{2} world"), StringInterpolation.new(["hello ".string, 1.int32, 2.int32, " world".string] of ASTNode)
+    it_parses %("hello \#{1}" \\\r\n "\#{2} world"), StringInterpolation.new(["hello ".string, 1.int32, 2.int32, " world".string] of ASTNode)
     it_parses "<<-HERE\nHello, mom! I am HERE.\nHER dress is beautiful.\nHE is OK.\n  HERESY\nHERE",
       "Hello, mom! I am HERE.\nHER dress is beautiful.\nHE is OK.\n  HERESY".string_interpolation
     it_parses "<<-HERE\n   One\n  Zero\n  HERE", " One\nZero".string_interpolation

--- a/spec/compiler/parser/parser_spec.cr
+++ b/spec/compiler/parser/parser_spec.cr
@@ -65,6 +65,7 @@ module Crystal
     it_parses %("foo"), "foo".string
     it_parses %(""), "".string
     it_parses %("hello \\\n     world"), "hello world".string
+    it_parses %("hello \\\r\n     world"), "hello world".string
 
     it_parses %(%Q{hello \\n world}), "hello \n world".string
     it_parses %(%q{hello \\n world}), "hello \\n world".string
@@ -1522,6 +1523,8 @@ module Crystal
 
     it_parses "1 \\\n + 2", Call.new(1.int32, "+", 2.int32)
     it_parses "1\\\n + 2", Call.new(1.int32, "+", 2.int32)
+    it_parses "1 \\\r\n + 2", Call.new(1.int32, "+", 2.int32)
+    it_parses "1\\\r\n + 2", Call.new(1.int32, "+", 2.int32)
 
     it_parses %("hello " \\\n "world"), StringLiteral.new("hello world")
     it_parses %("hello "\\\n"world"), StringLiteral.new("hello world")

--- a/src/compiler/crystal/syntax/lexer.cr
+++ b/src/compiler/crystal/syntax/lexer.cr
@@ -142,7 +142,9 @@ module Crystal
         consume_whitespace
         reset_regex_flags = false
       when '\\'
-        if next_char == '\n'
+        case next_char
+        when '\r', '\n'
+          handle_slash_r_slash_n_or_slash_n
           incr_line_number
           @token.passed_backslash_newline = true
           consume_whitespace
@@ -1373,7 +1375,9 @@ module Crystal
         when ' ', '\t'
           next_char
         when '\\'
-          if next_char == '\n'
+          case next_char
+          when '\r', '\n'
+            handle_slash_r_slash_n_or_slash_n
             next_char
             incr_line_number
             @token.passed_backslash_newline = true
@@ -1746,7 +1750,8 @@ module Crystal
                 buffer[0] = value
                 {1, 0}
               end
-            when '\n'
+            when '\r', '\n'
+              handle_slash_r_slash_n_or_slash_n
               incr_line_number
               @token.line_number = @line_number
 
@@ -1797,13 +1802,7 @@ module Crystal
           @token.value = "#"
         end
       when '\r', '\n'
-        is_slash_r = current_char == '\r'
-        if is_slash_r
-          if next_char != '\n'
-            raise "expecting '\\n' after '\\r'"
-          end
-        end
-
+        is_slash_r = handle_slash_r_slash_n_or_slash_n
         next_char
         incr_line_number 1
         @token.line_number = @line_number
@@ -2907,6 +2906,16 @@ module Crystal
       while (@token.type == :SPACE || @token.type == :NEWLINE || @token.type == :";")
         next_token
       end
+    end
+
+    def handle_slash_r_slash_n_or_slash_n
+      is_slash_r = current_char == '\r'
+      if is_slash_r
+        if next_char != '\n'
+          raise "expecting '\\n' after '\\r'"
+        end
+      end
+      is_slash_r
     end
 
     def unknown_token

